### ISSUE TITLE
Question: how to nest suites using inherited abstract classes?

### ISSUE THREAD

**Issue by blitzmann** [testdeck/testdeck#243]
Take the following example:

```
import { suite, test } from '@testdeck/mocha';
import { expect } from 'chai';

abstract class AbstractTestBase {

  protected abstract createNewInstance();

  @test
  inheritedTest() {

    expect(this.createNewInstance()).to.deep.equal({ 'foo': 'bar' });
  }
}

@suite
class ConcreteTests extends AbstractTestBase {

  protected createNewInstance() {

    return { 'foo': 'bar' };
  }

  @test
  specificTest() {

    expect(this.createNewInstance().foo).to.contain('bar');
  }
}
```

This will output 

```
...

  ConcreteTests
    ✓ specificTest
    ✓ inheritedTest
...
```

However, I would like it to output the following:
```
...
AbstractTestBase 
  ConcreteTests
    ✓ specificTest
    ✓ inheritedTest
...
```
I'm using inheritance to organize my test suites, but they have a certain hierarchy. At the root, I've got `BaseUnitTest`, and `BaseIntegrationTest`. From there, I have up to 3 levels of inheritance where I'm using `before()` in each one to set up some common functionality. However, as it currently stands, I get a full list of the concrete suites without any of that hierarchy printed out.

Is there any way that I can simulate this?  Thanks!

**Comment by blitzmann** [comment]
I guess what I'm asking is how to automatically nest these suites when they share a common base class. I don't want to go the route of https://testdeck.org/pages/guide/nested and stick all my concrete classes in the same describe - the file will be huge!

**Comment by silkentrance** [comment]
I have to ask why do you want/need such a feature?

Inheritance is meant for code/test reuse and not for nesting suites.

And while this could be done, I do not see any reason why it should be done.

**Comment by silkentrance** [comment]
As a side node: nesting itself is more about procedural generation of test suites within closures.

So if you have well defined unit test suites and integration test suites that do not require any external parameterisation from within closures, your best choice is to use inheritance.

You can combine this with procedural generation of test suites as well.

Otherwise, you might want to look at parameterised tests.

As for listing the abstract bases of a concrete suite in the output, I do not find any additional value in that and so I have to decline your feature request.

**Comment by silkentrance** [comment]
@blitzmann I need your input here

**Comment by blitzmann** [comment]
Hi @silkentrance , thanks for the response!

As stated, I'm using inheritance for both code reuse and to organize my test suites. Here's an example of what my inheritance structure might look like (also happens to be how the file system is laid out):

```
Integration
    Calculator
        Chaining
            suite1.test.ts
            suite2.test.ts
        Operator
            suite1.test.ts
            suite2.test.ts
            suite3.test.ts
        Modifiers
            suite1.test.ts
        Misc
            suite1.test.ts
            suite2.test.ts
    Container
    Effects
    Item
    Restrictions
    Simulation
    Stats
Unit
    ModBuilder
        Chaining
            suite1.test.ts
```

So, for each level, I have a class that the levels underneath it inherit from. So my `ChainingSuite1` concrete class suite inherits from abstract `ChainingTest`. which inherits from abstract `CalculatorTest` class, which inherits from abstract `IntegrationTest`. Most of these abstract classes adds additional setup to it's group of test suites by using `before()` (so, my calculator's `ChainingTest` class runs basic setup provided for all of it's test suites, using both it's own and inherited `before()` functions).

Also note that all the other levels underneath `Integration` will also have groups of tests, all ending in 1 or more suites, I just didn't feel like typing them out 😛 But there's over 1500 tests that will eventually be part of this.

I would like for an option for a visual representation of this structure. Currently, the suites run without any hierarchy, making it difficult to visually see whats going on if the concrete class doesn't have a very descriptive name. Lets say my concrete classes look like this:

`ChainingSuite1` under Integration > Calculator > Chaining
`ChainingSuite2` under Integration > Calculator > Chaining
`ChainingSuite1` under Unit > ModBuilder> Chaining

Since I have two with the name `ChainingSuite1`, I don't have any way to differentiate it. Right now, I would have to either name my classes very descriptively, or use `@suite`'s name option, both of which would require me to maintain the names correctly (if the hierarchy ever changes, will have to change everything associated with it). With potentially 100+ eventual test suites, there's bound to be some that have similar names.  I just figured having an option to use nesting for each inherited class might my a pretty cool win that would take care of this use case.

Admittedly, I'm new to node testing, and testing in general, so I'm not 100% surefooted on when you should nest suites, and don't know much about automatically generating test code and whatnot. I just noticed that nested suited gave me visually what I was looking for, but it's not practical to nest them manually like the docs suggest (the file would be huge.)

You're option, ie `INHERITANCE_NAMING`, would also work. I think it would look nicer for the tests to be nested via using the official describe blocks, but this would absolutely alleviate the issue of not being able to identify at a glance which suite is being run without maintaining more descriptive names manually. If this is something that could be included, that would be nice 😄 

Also, this is a great library, thank you for maintaining it. I come from a Python background, so when I started looking at node testing and realized that it in general wasn't OOP, I found this which has saved me. 😃

**Comment by silkentrance** [comment]
@blitzmann I also come from a Java/Python background, however, what you are asking for is basically a different reporting scheme.

Having abstract bases being reported in the test result output is actually a no-no, since the abstract bases will be inherited by the concrete suites.

A no-no since the abstract bases are not suites themselves, instead, they will provide derived concrete suites with test methods shared across multiple such concrete suites.

And this is not any different to JUNIT or PYTEST or any similar testing framework.

**Comment by silkentrance** [comment]
@blitzmann I also have a lot of experience with Python/Java, but never, ever I had encountered the need for reporting inheritance chains in my overall test reports.

So, no, we will not be adding such a feature. If you need such extraneously expressive reports, feel free to use the already available naming and nesting features.

**Comment by blitzmann** [comment]
> never, ever I had encountered the need for reporting inheritance chains in my overall test reports.

It's not so much that need to report the inheritance chain, but more to report the structure and organization of the files. I just figured using the inheritance would be an easy / sensible way of accomplishing this.

It's my understanding that a `describe` is used to to create a block of tests, which is useful when running `before` / `beforeEach`.

So, my tests in plain mocha would look something like this:

```
describe('Integration Tests', function () {
    beforeEach(() => { }) // Running set up for all integration tests
    describe('Calculation Tests', function () {
        beforeEach(() => { }) // Running set up for all calculation tests
        describe('Chaining Tests', function () {
            beforeEach(() => { }) // Running set up for all Chaining tests
            it('test 1', () => { });
            it('test 2', () => { });
            it('test 3', () => { });
            it('test 4', () => { });
        })
    });
});

```

The problem with this is that, in the Integration Tests block, I've got 7 main categories of Integration Tests. Each one of them can have a number of other groups of suites, or suites themselves. Doing it the plain mocha way would have been an organizational nightmare, with effectively one file for all Integration tests including all it's nested structure, thousands of lines.

This is one of the reasons why I'm using testdeck - I could more easily organize the files how I needed, and use familiar OOP inheritance to solve the before functionality for each test

But I can also see your point. In your opinion then, what is my best option here? I do not wish to put all these classes in huge describe blocks for obvious reasons to give my reporting some much needed structure. Is there a trick here that I'm missing? Perhaps if I could call `describe` in each group class  in a static `before` so that it runs once for all tests that inherit from it? I'm not sure how to accomplish that tho.

And thanks for taking the time to here me out :)
